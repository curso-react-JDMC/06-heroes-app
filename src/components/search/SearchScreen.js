import React, { useMemo } from "react";
import { useNavigate, useLocation } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import { getHeroesByName } from "../../selectors/getHeroesByName";
import { HeroCard } from "../hero/HeroCard";
import queryString from "query-string";
export const SearchScreen = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const { q = "" } = queryString.parse(location.search);

  const initialForm = {
    searchText: q,
  };

  const [formValues, handleInputChange, reset] = useForm(initialForm);
  const { searchText } = formValues;

  const heroesFiltered = useMemo(() => getHeroesByName(q),[q]);
  const handleSearch = (e) => {
    e.preventDefault();
    //console.log(searchText);
    navigate(`?q=${searchText}`);
  };

  return (
    <div>
      <h1>Search</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Buscar</h4>
          <hr />
          <form onSubmit={handleSearch}>
            <input
              type="text"
              placeholder="Buscar héroe"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={handleInputChange}
            />
            <button type="submit" className="btn btn-outline-primary mt-1">
              Buscar...
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Resultados</h4>
          <hr />
          {q === "" ? (
            <div className="alert alert-info">Buscar un Héroe</div>
          ) : (
            heroesFiltered.length === 0 && (
              <div className="alert alert-danger">Héroe no encontrado</div>
            )
          )}
          {heroesFiltered.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </div>
  );
};
