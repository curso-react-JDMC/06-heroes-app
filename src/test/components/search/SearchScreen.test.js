import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

const mockNavigate = jest.fn();
jest.mock('react-router-dom', ()=>({
    ...jest.requireActual('react-router-dom'),
    useNavigate:()=> mockNavigate
}))

describe('pruebas en SearchScreen', () => { 
    test('debe mostrarse correctamente con valores por defecto', () => { 
        const wrapper = mount(
        <MemoryRouter initialEntries={['/search']}>
            <SearchScreen/>
        </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text()).toBe('Buscar un Héroe');
    })
    test('debe mostrar a batman y el input con el valor del query String', () => { 
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
            <SearchScreen/>
        </MemoryRouter>
        )
        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
    })

    test('debe mostrar un error si no se encuentra el heroe', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
            <SearchScreen/>
        </MemoryRouter>
        )
        expect(wrapper.find('.alert-danger').text().trim()).toBe('Héroe no encontrado');
    })
    
    test('debe llamar el navigate a la nueva pantalla', () => { 
        const wrapper = mount(
        <MemoryRouter initialEntries={['/search']}>
            <SearchScreen/>
        </MemoryRouter>
        )
        wrapper.find('input').simulate('change',{
            target: {
                name:'searchText',
                value:'batman'
            }
        });
        wrapper.find('form').prop('onSubmit')({
            preventDefault:()=>{}
        })
        expect(mockNavigate).toHaveBeenCalledWith('?q=batman');
    })
})