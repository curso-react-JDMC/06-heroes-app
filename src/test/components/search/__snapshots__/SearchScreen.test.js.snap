// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en SearchScreen debe mostrar a batman y el input con el valor del query String 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar héroe"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit"
        >
          Buscar...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutter"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Batman"
                className="card-img-top"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                   
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Más..
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`pruebas en SearchScreen debe mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar héroe"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit"
        >
          Buscar...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Buscar un Héroe
      </div>
    </div>
  </div>
</div>
`;
